### DISCUSSION TITLE
Queries related to discretization and modelling.

### DISCUSSION THREAD

**Discussion by sarthaks20** [peridigm/peridigm#315] [Q&A] [answered]
Hey everyone!

I have few queries regarding discretization and modelling. I would be really grateful if anyone could answer.

1. Can I use the mesh I created in ABAQUS FEA by converting it to Genesis (.g) file or any other supported format? If yes, how?
2. What is the difference between .g and .e files?
3. How to model a rigid body in Peridigm?
4. How to use parallel computing by using CPUs and GPUs simultaneously?

Thank you!

**Comment by djlittl** [comment, marked as answer]
1.  It depends.  Peridigm can read 4-node tet elements and 8-node hex elements and automatically convert them to the needed meshless discretization.  So if the ABAQUS mesh is comprised of hex and tet elements, you should be able to get this to work.  If there are shell elements, or similar, you'd have to come up with a scheme for converting those into something Peridigm can handle.  If the ABAQUS mesh contains only tets and hexes, then as you implied you just need to convert the file format.  I'm not aware of any tools that can do this automatically, but I could be wrong about that.  To do it manually, you could write a Python script that reads the ABAQUS mesh into Python data structures, and then you could write the data to a .g file using the exodus.py module.  There are some Python scripts in the Peridigm repository that use exodus.py, you could have a look at those to get a sense of how it works.  Peridigm can also read a very simple text file format for meshes, but exodus/genesis is preferred for efficiency.
2. I think technically there might not be any difference.  But by convention, .g denotes an input file (g = genesis) and .e denotes an output file (e = exodus).  From a file format point of view, I think a .g file is a .e file with no data fields, just the nodal coordinates and element connectivity.  Note that Peridigm uses the exodus SPHERE element type to store the meshless discretization, so the connectivity array is trivial.
3. There is no functionality for a rigid body in Peridigm.
4. There is no functionality for utilizing GPUs in Peridigm.  Parallel computing on CPUs is achieved via MPI, and it works well.  Utilizing GPUs would require the Kokkos package from Trilinos (strongly preferred) or use of a GPU-specific programming language like CUDA.

**Comment by oldninja** [comment]
@djlittl :

3.  Do you have any recommendations on how to code a rigid body in Peridigm? 
4. Do you anticipate a performance improvement (w.r.t. simulation time) if we add GPU acceleration? Are there any benchmarks?

**Comment by djlittl** [comment]
1. I don't have any experience with this, so unfortunately I don't have any recommendations.  I assume we could borrow heavily from the FEM community on this.
2. I do think there would be a performance gain, but I'm not convinced it would be worth the cost in terms of developer effort or increased code complexity.  @diehlpk would have better insight than me.  I personally would prioritize refactoring contact and improving the build system and build documentation over going after performance improvements with GPUs.  But I get it that the GPU work would be more fun :) and would easier to sell as a R&D project to a sponsor.

**Comment by djlittl** [comment]
Unrelated topic:  I recently did a literature review on groups that use Peridigm, and I came across many of the folks that are active on this site.  But I could not figure out who @oldninja is.  I also could not find a way to message someone directly via GitHub (I think this is by design).  @oldninja could you send me an email so we can connect in the real world?  And more broadly it would be great to meet folks who use Peridigm at the USNCCM conference, which we are hosting this year in Albuquerque.

**Comment by diehlpk** [comment]
Yes, I would agree that improving the documentation and build system would be better for the community. I am not sure if we will gain many benefits using GPUs for the implicit part, since we would need to solve a large matrix on GPUs. We used Kokkos to rewrite a large code base, and this was a major effort.
